affinity/cpp-23/dXXXXr0.md

```diff
@@ -0,0 +1,295 @@
+An *execution resource* represents an abstraction of a hardware or software
+layer that guarantees a particular set of affinity properties, where the level
+of abstraction is implementation-defined. An implementation is permitted to
+migrate any underlying resources providing it guarantees the affinity properties
+remain consistent. This allows freedom for the implementor but also consistency
+for users.
+
+If an *execution resource* is valid, then it must always point to the same
+underlying thing. For example, a *resource* cannot first point to one CPU core,
+and then suddenly point to a different CPU core. An *execution context* can thus
+rely on properties like binding of operating system threads to CPU cores.
+However, the "thing" to which an *execution resource* points may be a dynamic,
+possibly a software-managed pool of hardware. Here are three examples of this
+phenomenon:
+
+   1. The "hardware" may actually be a virtual machine (VM). At any point, the
+   VM may pause, migrate to different physical hardware, and resume. If the VM
+   presents the same virtual hardware before and after the migration, then the
+   *resources* that an application running on the VM sees should not change.
+   2. The OS may maintain a pool of a varying number of CPU cores as a shared
+   resource among different user-level processes. When a process stops using the
+   resource, the OS may reclaim cores. It may make sense to present this pool as
+   an *execution resource*.
+   3. A low-level device driver on a laptop may switch between a "discrete" GPU
+   and an "integrated" GPU, depending on utilization and power constraints. If
+   the two GPUs have the same instruction set and can access the same memory, it
+   may make sense to present them as a "virtualized" single *execution
+   resource*.
+
+In summary, an *execution resource* either identifies a thing uniquely, or
+harmlessly points to nothing.
+
+### Property group `adjacency`
+
+The `adjacency_t` properties are a group of mutually exclusive behavioral
+properties which provide a hint to an executor, via an *execution policy* which
+describes the implication of adjacent work-items of the callable of a bulk
+algorithm, allowing the implementation to choose the appropriate action.
+
+*Locality distance* specifies an implementation-defined metric for
+measuring the relative affinity between *execution resources* whereby *execution
+resources* with a lower *locality distance* are likely to have similar latency
+in memory access operations, for a given memory location.
+
+> [*Note:* For example, two hardware threads on the same CPU core would have a
+lower *locality distance* whereas two hardware threads on different CPU cores or
+in different NUMA regions would have a higher *locality distance*. *--end note*]
+
+> [*Note:* An alternative term of art for *locality distance* could be *locality
+interference*. *--end note*]
+
+When a bulk algorithm is invoked with a callable `f`, a size `s` and using a
```

Comment:
A thought came to me when writing this, in previous papers we had wording which suggested implementations should aim to map work-items to executions resources in a deterministic way (given the same size). Should we introduce this for the `adjacency` property as a hint or should we leave this as a quality of implementation detail or a future property?

Reply:
@AerialMantis wrote:
> Should we introduce this for the `adjacency` property as a hint or should we leave this as a quality of implementation detail or a future property?

I'd like to see it, not as a quality of implementation detail, but either as default behavior or as something I can request (either a hint or property).  Even if the default is a deterministic mapping, it would be nice to be able to relax that by request.  Existing systems (e.g., Kokkos and OpenMP) have both static (deterministic) and dynamic (nondeterministic) options for assigning chunks of work to execution resources, and I think users would want us to expose that.

Reply:
I agree, I think having a property that allows you to specify a level of determinism in the mapping of work-items in bulk algorithms. I can make determinism the default for now as this is most beneficial for the adjacency property.

Reply:
I've added this as a non-normative note, since the properties are not prescriptive it's hard to require a particular mapping, how does this sound?

Subsequent invocations of a bulk algorithm with the same size `s` and the same policy `p`, scheduling on the same executor `e` should aim to map work-items to *execution resources* consistently with previous invocations.